xds/internal/balancer/edsbalancer/eds.go

```diff
@@ -103,340 +99,253 @@ func (bb) ParseConfig(c json.RawMessage) (serviceconfig.LoadBalancingConfig, err
-//
-// If the error is resource-not-found:
-// - If it's from CDS balancer (shows as a resolver error), it means LDS or CDS
-// resources were removed. The EDS watch should be canceled.
-// - If it's from xds client, it means EDS resource were removed. The EDS
-// watcher should keep watching.
-// In both cases, the sub-balancers will be closed, and the future picks will
-// fail.
-func (b *edsBalancer) handleErrorFromUpdate(err error, fromParent bool) {
-	b.logger.Warningf("Received error: %v", err)
-	if xdsclient.ErrType(err) == xdsclient.ErrorTypeResourceNotFound {
-		if fromParent {
-			// This is an error from the parent ClientConn (can be the parent
-			// CDS balancer), and is a resource-not-found error. This means the
-			// resource (can be either LDS or CDS) was removed. Stop the EDS
-			// watch.
-			b.cancelWatch()
-		}
-		b.edsImpl.handleEDSResponse(xdsclient.EndpointsUpdate{})
+// handleClientConnUpdate handles a ClientConnUpdate received from gRPC. Good
+// updates lead to registration of an EDS watch. Updates with error lead to
+// cancellation of existing watch and propagation of the same error to the
+// child balancer.
+func (b *edsBalancer) handleClientConnUpdate(update *ccUpdate) {
+	// We first handle errors, if any, and then proceed with handling the
+	// update, only if the status quo has changed.
+	if err := update.err; err != nil {
+		b.handleErrorFromUpdate(err, true)
 	}
-}
 
-func (b *edsBalancer) handleGRPCUpdate(update interface{}) {
-	switch u := update.(type) {
-	case *subConnStateUpdate:
-		b.edsImpl.handleSubConnStateChange(u.sc, u.state.ConnectivityState)
-	case *balancer.ClientConnState:
-		b.logger.Infof("Received update from resolver, balancer config: %+v", pretty.ToJSON(u.BalancerConfig))
-		cfg, _ := u.BalancerConfig.(*EDSConfig)
-		if cfg == nil {
-			// service config parsing failed. should never happen.
-			return
-		}
-
-		if err := b.handleServiceConfigUpdate(cfg); err != nil {
-			b.logger.Warningf("failed to update xDS client: %v", err)
-		}
-
-		b.edsImpl.updateServiceRequestsConfig(cfg.ClusterName, cfg.MaxConcurrentRequests)
-
-		// We will update the edsImpl with the new child policy, if we got a
-		// different one.
-		if !cmp.Equal(cfg.ChildPolicy, b.config.ChildPolicy, cmpopts.EquateEmpty()) {
-			if cfg.ChildPolicy != nil {
-				b.edsImpl.handleChildPolicy(cfg.ChildPolicy.Name, cfg.ChildPolicy.Config)
-			} else {
-				b.edsImpl.handleChildPolicy(roundrobin.Name, nil)
-			}
-		}
-		b.config = cfg
-	case error:
-		b.handleErrorFromUpdate(u, true)
-	default:
-		// unreachable path
-		b.logger.Errorf("wrong update type: %T", update)
+	b.logger.Infof("Receive update from resolver, balancer config: %+v", update.state.BalancerConfig)
+	cfg, _ := update.state.BalancerConfig.(*EDSConfig)
+	if cfg == nil {
+		b.logger.Warningf("xds: unexpected LoadBalancingConfig type: %T", update.state.BalancerConfig)
+		// service config parsing failed. should never happen.
```

Comment:
Nit: Consider moving this comment to the line above the log message. And since this is a full sentence, treat it as one, and capitalize the first letter.

`should never happen`
Why? Could you please clarify that in the comment.